File: app/src/main/java/com/novel/read/adapter/BookAdapter.java
package com.novel.read.adapter;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.novel.read.R;
import com.novel.read.activity.NovelReadActivity;
import com.novel.read.model.db.CollBookBean;
import com.novel.read.utlis.GlideImageLoader;

import java.util.ArrayList;
import java.util.List;

public class BookAdapter extends RecyclerView.Adapter {

    private List<CollBookBean> mList;
    private Context mContext;
    private final int VALUE_ITEM = 100; //正常item
    private final int EMPTY_ITEM = 101; //空白item
    protected OnItemClickListener mClickListener;
    private boolean mEdit;

    public BookAdapter(List<CollBookBean> mList) {
        this.mList = mList;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        if (mContext == null) {
            mContext = viewGroup.getContext();
        }
        View view;
        if (i == VALUE_ITEM) {
            view = LayoutInflater.from(mContext).inflate(R.layout.rlv_item_book, viewGroup, false);
            return new ViewHolder(view);
        } else if (i == EMPTY_ITEM) {
            view = LayoutInflater.from(mContext).inflate(R.layout.rlv_empty_add_book, viewGroup, false);
            return new EmptyHolder(view);
        }
        return null;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int i) {
        if (viewHolder instanceof ViewHolder) {
            if (mEdit) { //编辑模式
                ((ViewHolder) viewHolder).mIvCheck.setSelected(mList.get(i).isSelect());
                ((ViewHolder) viewHolder).mIvCheck.setVisibility(View.VISIBLE);
                ((ViewHolder) viewHolder).mTvBookName.setText(mList.get(i).getTitle());
                ((ViewHolder) viewHolder).mTvBookAuthor.setText(mList.get(i).getAuthor());
                if (mList.get(i).isUpdate()) {
                    ((ViewHolder) viewHolder).mIvGeng.setVisibility(View.VISIBLE);
                } else {
                    ((ViewHolder) viewHolder).mIvGeng.setVisibility(View.GONE);
                }
                GlideImageLoader.displayCornerImage(mContext, mList.get(i).getCover(), ((ViewHolder) viewHolder).mIvBook);
                viewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        ((ViewHolder) viewHolder).mIvCheck.setSelected(!mList.get(i).isSelect());
                        mList.get(i).setSelect(!mList.get(i).isSelect());
                    }
                });
            } else { //正常模式
                ((ViewHolder) viewHolder).mIvCheck.setVisibility(View.GONE);
                if (mList.size() == i) { //最后的条目
                    ((ViewHolder) viewHolder).mTvBookName.setText("");
                    ((ViewHolder) viewHolder).mTvBookAuthor.setText("");
                    ((ViewHolder) viewHolder).mIvBook.setImageResource(R.drawable.ic_book_add);
                    ((ViewHolder) viewHolder).mIvGeng.setVisibility(View.GONE);
                    viewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            mClickListener.onItemClick(view, i);
                        }
                    });
                } else {
                    ((ViewHolder) viewHolder).mTvBookName.setText(mList.get(i).getTitle());
                    ((ViewHolder) viewHolder).mTvBookAuthor.setText(mList.get(i).getAuthor());
                    if (mList.get(i).isUpdate()) {
                        ((ViewHolder) viewHolder).mIvGeng.setVisibility(View.VISIBLE);
                    } else {
                        ((ViewHolder) viewHolder).mIvGeng.setVisibility(View.GONE);
                    }
                    GlideImageLoader.displayCornerImage(mContext, mList.get(i).getCover(), ((ViewHolder) viewHolder).mIvBook);
                    viewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            Intent intent = new Intent(mContext, NovelReadActivity.class);
                            intent.putExtra(NovelReadActivity.EXTRA_IS_COLLECTED, true);
                            intent.putExtra(NovelReadActivity.EXTRA_COLL_BOOK, mList.get(i));
                            mContext.startActivity(intent);
                        }
                    });
                }
            }
        } else if (viewHolder instanceof EmptyHolder) { //空条目
            ((EmptyHolder) viewHolder).mBtnAdd.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    mClickListener.onItemClick(view, i);
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        if (mList.size() == 0) {
            return 1;
        }
        if (mEdit) {
            return mList.size();
        } else {
            return mList.size() + 1;
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (mList == null || mList.size() == 0) {
            return EMPTY_ITEM;
        } else {
            return VALUE_ITEM;
        }

    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        ImageView mIvBook;
        TextView mTvBookName, mTvBookAuthor;
        ImageView mIvCheck;
        ImageView mIvGeng;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            mIvBook = itemView.findViewById(R.id.iv_book);
            mTvBookName = itemView.findViewById(R.id.tv_book_name);
            mTvBookAuthor = itemView.findViewById(R.id.tv_book_author);
            mIvCheck = itemView.findViewById(R.id.iv_check);
            mIvGeng = itemView.findViewById(R.id.iv_geng);
        }
    }

    static class EmptyHolder extends RecyclerView.ViewHolder {
        Button mBtnAdd;

        public EmptyHolder(@NonNull View itemView) {
            super(itemView);
            mBtnAdd = itemView.findViewById(R.id.btn_add);
        }
    }

    public void setOnItemClickListener(OnItemClickListener mListener) {
        this.mClickListener = mListener;
    }


    public interface OnItemClickListener {
        void onItemClick(View view, int pos);
    }

    public void setEdit(boolean edit) {  //开启编辑模式
        mEdit = edit;
        notifyDataSetChanged();
    }

    public List<CollBookBean> getSelectList() {
        List<CollBookBean> collBookBeans = new ArrayList<>();
        for (int i = 0; i < mList.size(); i++) {
            if (mList.get(i).isSelect()) {
                collBookBeans.add(mList.get(i));
            }
        }
        return collBookBeans;
    }

}

File: app/src/main/java/com/novel/read/adapter/BookAdapter.kt
package com.novel.read.adapter

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

import com.novel.read.R
import com.novel.read.activity.NovelReadActivity
import com.novel.read.model.db.CollBookBean
import com.novel.read.utlis.GlideImageLoader

import java.util.ArrayList

class BookAdapter(private val mList: List<CollBookBean>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var mContext: Context? = null

    private var mClickListener: OnItemClickListener? = null
    private var mEdit: Boolean = false

    val selectList: List<CollBookBean>
        get() {
            val collBookBeans = ArrayList<CollBookBean>()
            for (i in mList.indices) {
                if (mList[i].isSelect) {
                    collBookBeans.add(mList[i])
                }
            }
            return collBookBeans
        }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): RecyclerView.ViewHolder {
        if (mContext == null) {
            mContext = viewGroup.context
        }
        val view: View
        if (i == VALUE_ITEM) {
            view = LayoutInflater.from(mContext).inflate(R.layout.rlv_item_book, viewGroup, false)
            return ViewHolder(view)
        } else if (i == EMPTY_ITEM) {
            view = LayoutInflater.from(mContext).inflate(R.layout.rlv_empty_add_book, viewGroup, false)
            return EmptyHolder(view)
        }
        throw IllegalArgumentException()
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, i: Int) {
        if (viewHolder is ViewHolder) {
            if (mEdit) { //编辑模式
                viewHolder.mIvCheck.isSelected = mList[i].isSelect
                viewHolder.mIvCheck.visibility = View.VISIBLE
                viewHolder.mTvBookName.text = mList[i].title
                viewHolder.mTvBookAuthor.text = mList[i].author
                if (mList[i].isUpdate) {
                    viewHolder.mIvGeng.visibility = View.VISIBLE
                } else {
                    viewHolder.mIvGeng.visibility = View.GONE
                }
                GlideImageLoader.displayCornerImage(mContext!!, mList[i].cover, viewHolder.mIvBook)
                viewHolder.itemView.setOnClickListener {
                    viewHolder.mIvCheck.isSelected = !mList[i].isSelect
                    mList[i].isSelect = !mList[i].isSelect
                }
            } else { //正常模式
                viewHolder.mIvCheck.visibility = View.GONE
                if (mList.size == i) { //最后的条目
                    viewHolder.mTvBookName.text = ""
                    viewHolder.mTvBookAuthor.text = ""
                    viewHolder.mIvBook.setImageResource(R.drawable.ic_book_add)
                    viewHolder.mIvGeng.visibility = View.GONE
                    viewHolder.itemView.setOnClickListener { view ->
                        mClickListener!!.onItemClick(
                            view,
                            i
                        )
                    }
                } else {
                    viewHolder.mTvBookName.text = mList[i].title
                    viewHolder.mTvBookAuthor.text = mList[i].author
                    if (mList[i].isUpdate) {
                        viewHolder.mIvGeng.visibility = View.VISIBLE
                    } else {
                        viewHolder.mIvGeng.visibility = View.GONE
                    }
                    GlideImageLoader.displayCornerImage(
                        mContext!!,
                        mList[i].cover,
                        viewHolder.mIvBook
                    )
                    viewHolder.itemView.setOnClickListener {
                        val intent = Intent(mContext, NovelReadActivity::class.java)
                        intent.putExtra(NovelReadActivity.EXTRA_IS_COLLECTED, true)
                        intent.putExtra(NovelReadActivity.EXTRA_COLL_BOOK, mList[i])
                        mContext!!.startActivity(intent)
                    }
                }
            }
        } else if (viewHolder is EmptyHolder) { //空条目
            viewHolder.mBtnAdd.setOnClickListener { view -> mClickListener!!.onItemClick(view, i) }
        }
    }

    override fun getItemCount(): Int {
        if (mList.isEmpty()) {
            return 1
        }
        return if (mEdit) {
            mList.size
        } else {
            mList.size + 1
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (mList.isEmpty()) {
            EMPTY_ITEM
        } else {
            VALUE_ITEM
        }

    }

    internal class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var mIvBook: ImageView = itemView.findViewById(R.id.iv_book)
        var mTvBookName: TextView = itemView.findViewById(R.id.tv_book_name)
        var mTvBookAuthor: TextView = itemView.findViewById(R.id.tv_book_author)
        var mIvCheck: ImageView = itemView.findViewById(R.id.iv_check)
        var mIvGeng: ImageView = itemView.findViewById(R.id.iv_geng)

    }

    internal class EmptyHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var mBtnAdd: Button = itemView.findViewById(R.id.btn_add)

    }

    fun setOnItemClickListener(mListener: OnItemClickListener) {
        this.mClickListener = mListener
    }


    interface OnItemClickListener {
        fun onItemClick(view: View, pos: Int)
    }

    fun setEdit(edit: Boolean) {  //开启编辑模式
        mEdit = edit
        notifyDataSetChanged()
    }


    companion object {

        private const val VALUE_ITEM = 100 //正常item

        private const val EMPTY_ITEM = 101 //空白item

    }

}
